x/cdp/keeper/interest.go

```diff
@@ -86,129 +110,41 @@ func CalculateInterestFactor(perSecondInterestRate sdk.Dec, secondsElapsed sdk.I
-	feesAccumulated := (sdk.NewDecFromInt(principal.Amount).Mul(accumulator)).Sub(sdk.NewDecFromInt(principal.Amount))
-	newFees := sdk.NewCoin(principal.Denom, feesAccumulated.TruncateInt())
-	return newFees
-}
-
-// IncrementTotalPrincipal increments the total amount of debt that has been drawn with that collateral type
-func (k Keeper) IncrementTotalPrincipal(ctx sdk.Context, collateralType string, principal sdk.Coin) {
-	total := k.GetTotalPrincipal(ctx, collateralType, principal.Denom)
-	total = total.Add(principal.Amount)
-	k.SetTotalPrincipal(ctx, collateralType, principal.Denom, total)
-}
+		k.SetCDP(ctx, cdp)
+	}
+	cdpInterestFactor := sdk.OneDec().Add((globalInterestFactor.Sub(cdp.InterestFactor)))
+	if cdpInterestFactor.Equal(sdk.OneDec()) {
+		return cdp
+	}
 
-// DecrementTotalPrincipal decrements the total amount of debt that has been drawn for a particular collateral type
-func (k Keeper) DecrementTotalPrincipal(ctx sdk.Context, collateralType string, principal sdk.Coin) {
-	total := k.GetTotalPrincipal(ctx, collateralType, principal.Denom)
-	// NOTE: negative total principal can happen in tests due to rounding errors
-	// in fee calculation
-	total = sdk.MaxInt(total.Sub(principal.Amount), sdk.ZeroInt())
-	k.SetTotalPrincipal(ctx, collateralType, principal.Denom, total)
-}
+	accumulatedInterest := cdp.GetTotalPrincipal().Amount.ToDec().Mul(cdpInterestFactor).RoundInt().Sub(cdp.GetTotalPrincipal().Amount)
 
-// GetTotalPrincipal returns the total amount of principal that has been drawn for a particular collateral
-func (k Keeper) GetTotalPrincipal(ctx sdk.Context, collateralType, principalDenom string) (total sdk.Int) {
-	store := prefix.NewStore(ctx.KVStore(k.key), types.PrincipalKeyPrefix)
-	bz := store.Get([]byte(collateralType + principalDenom))
-	if bz == nil {
-		k.SetTotalPrincipal(ctx, collateralType, principalDenom, sdk.ZeroInt())
-		return sdk.ZeroInt()
-	}
-	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &total)
-	return total
+	cdp.AccumulatedFees = cdp.AccumulatedFees.Add(sdk.NewCoin(cdp.AccumulatedFees.Denom, accumulatedInterest))
+	cdp.FeesUpdated = ctx.BlockTime()
+	cdp.InterestFactor = globalInterestFactor
+	collateralToDebtRatio := k.CalculateCollateralToDebtRatio(ctx, cdp.Collateral, cdp.Type, cdp.GetTotalPrincipal())
+	k.SetCdpAndCollateralRatioIndex(ctx, cdp, collateralToDebtRatio)
+	return cdp
 }
 
-// SetTotalPrincipal sets the total amount of principal that has been drawn for the input collateral
-func (k Keeper) SetTotalPrincipal(ctx sdk.Context, collateralType, principalDenom string, total sdk.Int) {
-	store := prefix.NewStore(ctx.KVStore(k.key), types.PrincipalKeyPrefix)
-	_, found := k.GetCollateralTypePrefix(ctx, collateralType)
+// CalculateNewInterest returns the amount of interest that has accrued to the cdp since its interest was last synchronized
+func (k Keeper) CalculateNewInterest(ctx sdk.Context, cdp types.CDP) sdk.Coin {
```

Comment:
Looks like this method is only used in `interest_test.go` and `operations.go` when we need to query a CDP's outstanding interest as CDP queries directly to the store now return out-of-date information. The front-end is going to need to display up-to-date CDP info as well; we should call this function at the start of `LoadAugmentedCDP()` and add the new fees accumulated.

Reply:
19e93cc